readme.md
# @brixtol/pjax

A blazing fast, lightweight (8.2kb gzipped) and feature full new generation pjax solution. This pjax variation supports advanced pre-fetching capabilities, multiple fragment replacements and employs a snapshot caching engine that prevents subsequent requests from occurring.

### Features

- Simple and painless integration.
- Pre-fetching capabilities using hover, intersect and proximity triggers.
- Snapshot caching engine and per-page state control.
- Powerful pub/sub event driven lifecycle triggers.
- Provides a client side DOM hydration approach.
- Supports targeted append and prepend fragment replacements.
- Handles inlined and external script evaluation (asynchronously).
- Couples perfectly with [stimulus.js](https://stimulusjs.org/).
- Attribute driven programmatic control.

### Demo

We are using this module live on our [webshop](https://brixtoltextiles.com).

### Why?

The landscape of pjax based solution has become rather scarce. The current bread winners tend to offer the same thing or for our use case were either over or under engineered. This pjax variation couples together various techniques found to be the most effective in enhancing the performance of SSR rendered web application that fetch pages over the wire.

# Table of Contents

1. [Install](#install)
2. [Recommendations](#recommendations)
3. [Usage](#usage)
4. [Options](#options)
5. [Real World](#real-world)
6. [Lifecycle Events](#lifecycle-events)
7. [Methods](#methods)
8. [Attributes](#attributes)
9. [State](#state)
10. [How it works](#how-it-works)
11. [Contributing](#contributing)
12. [Acknowledgements](#acknowledgements)

# Install

This module is distributed as ESM and designed to work in the browser environment.

### pnpm

```cli
pnpm add @brixtol/pjax
```

_Because [pnpm](https://pnpm.js.org/en/cli/install) is dope and does dope shit._

### Yarn

```cli
yarn add @brixtol/pjax
```

_Stop using Yarn, it sucks. Choose [pnpm](https://pnpm.js.org/en/cli/install) and emancipate yourself._

### npm

```cli
npm i @brixtol/pjax
```

_Okay, Boomer..._

### cdn

```
https://unpkg.com/@brixtol/pjax
```

_Be as you are.._

# Recommendations

In order to get the most of this module below are a few recommendations developers should consider when leveraging it within in their projects.

### Pre-fetching

The pre-fetching capabilities this Pjax variation provides can drastically improve the speed of rendering. When used correctly pages will load instantaneously between navigations. By default, the pre-fetching features are opt-in and require attribute annotation but you can customize how, when and where pjax should execute a pre-fetch.

### Stimulus

This module was developed as a replacement for [turbo](https://github.com/hotwired/turbo) so leveraging it together with [stimulus.js](https://stimulusjs.org/) is the preferred usage. Stimulus is a very simple framework and when working with SSR projects it helps alleviate the complications developers tend to face. The reason one would choose this project over Turbo comes to performance as this module is much faster and smaller than turbo.

### Minification

By default, all fetched pages are stored in memory so for every request the HTML dom string response is saved to cache. The smaller your HTML pages the more performant the rendering engine will be. In addition to minification it is generally good practice to consider using semantic HTML5 as much as possible so as to negate the amount of markup pages require.

### JavaScript

Inline javascript contained within `<script></script>` tags should be placed in your documents `<head>` element. Script tags will be evaluated and executed on a per-page basis. You can prevent scripts from being re-evaluated by attributing them with `data-pjax-eval="false"` attributes. When script tags contain this attribute they will be executed only once and never again after that. Scripts which reference JavaScript files, ie: `<script src="*"></script>` are evaluated only onceonce and never again unless attributed with a `data-pjax-eval="true"` attribute.

The best possible approach is to initialize JavaScript with `pjax.on('load', () => {})` method event. This way you can be sure it will load between navigations and improve performance.

# Usage

To initialize, call `pjax.connect()` in your bundle preferably before anything else is loaded. By default, the entire `<body>` fragment is replaced upon each navigation. You should define a set of `targets[]` whose inner contents change on a per-page basis for optimal performance and consider leveraging the pre-fetching capabilities.

> The typings provided in this package will describe each option in good detail, below are the defaults and all options are optional.

```js
import * as pjax from '@brixtol/pjax';

pjax.connect({
  targets: ['body'],
  schema: 'pjax',
  timeout: 30000,
  poll: 15,
  async: true,
  cache: true,
  reverse: true,
  persist: false,
  limit: 50,
  preload: null,
  hover: {
    trigger: 'attribute',
    threshold: 250
  },
  intersect: {
    rootMargin: '',
    threshold: 0
  },
  proximity: {
    distance: 85,
    throttle: 500,
    threshold: 250
  },
  progress: {
    minimum: 0.08,
    easing: 'linear',
    speed: 200,
    trickle: true,
    threshold: 500,
    trickleSpeed: 200
  }
});
```

# Options

### `targets`

Define page fragment targets which are expected to change on a per-page basis. By default, this pjax module will replace the entire `<body>` fragment. It's best to define specific fragments.

**Type:** `string[]` <br>
**Default:** `['body']` <br>

### `schema`

By default, attribute identifiers use a `-pjax-` identifier. You can use a custom attribute identifier.

**Type:** `string` <br>
**Default:** `pjax` <br>

### `timeout`

Request polling limit is used when a request is already in transit. Request completion is checked every 10ms, by default this is set to `30000` which means requests will wait `30s` before being a new request is triggered.

### `poll`

Request polling limit is used when a request is already in transit. Request completion is checked every 10ms, by default this is set to `1000` which means requests will wait `1s` before being a new request is triggered.

**Type:** `number` <br>
**Default:** `1000` <br>

### `async`

Determine if page requests should be fetched asynchronously or synchronously. Setting this to `false` is not recommended.

**Type:** `boolean` <br>
**Default:** `true` <br>

### `cache`

Enable or Disable caching. Each page visit request is cached and used in subsequent visits to the same location. Setting this to `false` is discourage as all visits will be fetched over the network and `data-pjax-cache` attribute configs will be ignored.

> If `cache` is disabled then prefetches will be dispatched using HTML5 `<link>` prefetches, else when cache is enabled it uses XHR.

**Type:** `boolean` <br>
**Default:** `true` <br>

### `persist`

**NOT YET AVAILABLE**

The `persist` option can be used to restore cache into memory after a browser refresh has been triggered. When persisting cache a reference is maintained in session storage.

**Type:** `boolean` <br>
**Default:** `false` <br>

### `reverse`

Reverse caching. This will execute a pre-emptive fetch of the previous pages in the history stack when no snapshot exists in cache. Snapshots cache is purged when browser refresh occurs (unless `persist` is enabled) so when navigating backwards or pages will need to be re-fetched and this results in minor delays due to the refresh which was triggered.

**Type:** `boolean` <br>
**Default:** `true` <br>

### `limit`

Cache size limit. This pjax variation limits cache size to `50mb`and once it exceeds that limit, records will be removed starting from the earliest point of known cache entries.

**Type:** `number` <br>
**Default:** `50` <br>

### `hover`

Hover pre-fetching. You can disable hover pre-fetching by setting this to `false` which will prevent observers from executing and any `data-pjax-hover` attributes will be ignored. To use the default configurations you can set this to `true` or simply omit it.

**Type:** `boolean` or `object` <br>
**Default:** `{ trigger: 'attribute', threshold: 250 }` <br>

### `hover.trigger`

How hover prefetches should be triggered. By default this option is set to trigger only when `<a>` href link elements are attributed with a `data-pjax-hover` attribute. You can instruct pjax to execute pre-fetching on all `<a>` elements by setting this option to `href`. If you set the trigger to `href` you can annotate links you wish to exclude from prefetch with `data-pjax-hover="false"`.

**Type:** `string` <br>
**Accepts:** `attribute` or `href` <br>
**Default:** `attribute` <br>

### `hover.threshold`

Controls the fetch delay threshold. Requests will fire only when the mouse is both within range and the threshold time limit defined here has exceeded.

**Type:** `number` <br>
**Default:** `250` <br>

### `proximity`

Proximity pre-fetching allow for requests to be dispatched when the cursor is within a proximity range of a href link element. Coupling proximity with mouseover prefetches enable predicative fetching to occur, so a request will trigger before any interaction with a link element happens. To use default behavior, set this to `true` and all `<a>` annotated with a `data-pjax-proximity` attribute will be pre-fetched.

> Annotate any `<a>` links you wish to exclude from pre-fetching using the `data-pjax-proximity="false"`

**Type:** `boolean` or `object` <br>
**Default:** `{ distance: 75, throttle: 500, threshold: 250 }` <br>

### `proximity.distance`

The distance range the mouse should be within before the prefetch is triggered. You can optionally override this by assigning a number value to the proximity attribute. An href element using `data-pjax-proximity="50"` would inform Pjax to begin fetching when the mouse is within `50px` of the element.

**Type:** `number` <br>
**Default:** `75` <br>

### `proximity.throttle`

Controls the fetch delay threshold. Requests will fire only when the mouse is both within range and the threshold time limit defined here has exceeded.

**Type:** `number` <br>
**Default:** `250` <br>

### `proximity.threshold`

Controls the fetch delay threshold. Requests will fire only when the mouse has exceeded the range and the threshold time limit defined here has been exceeded.

**Type:** `number` <br>
**Default:** `250` <br>

### `intersect`

Intersection pre-fetching. Intersect pre-fetching leverages the [Intersection Observer](https://shorturl.at/drLW9) API to fire requests when elements become visible in viewport. You can disable intersect pre-fetching by setting this to `false`, otherwise you can customize the intersect fetching behavior. To use default behavior, set this to `true` and all elements annotated with with a `data-pjax-intersect` or `data-pjax-intersect="true"` attribute will be pre-fetched. You can annotate elements that contain href links or `<a>` elements directly.

> Annotate any `<a>` links you wish to exclude from intersection pre-fetching using the `data-pjax-intersect="false"`

**Type:** `boolean` or `object` <br>
**Default:** `{ rootMargin: '0px 0px 0px 0px', throttle: 0 }` <br>

### `intersect.rootMargin`

An offset rectangle applied to the root's href bounding box. The option is passed to the Intersection Observer.

**Type:** `string` <br>
**Default:** `0px 0px 0px 0px` <br>

### `intersect.throttle`

Throttle limit passed to the intersection observer instance.

**Type:** `number` <br>
**Default:** `500` <br>

# Real World

Below is a real world example you can use to better understand how this module works and how it can be applied to your web application. We are working on providing a live demonstration for more advanced use cases but the below example should give you a good understanding of how to leverage this module.

<details>
<summary>
<strong>Example</strong>
</summary>

The first thing we want to do is make a connection with Pjax. In your JavaScript bundle we need to initialize it (connect). Our example web application has 3 pages, the **home** page, **about** page and **contact** page. We are going to instruct pjax to replace the `<nav>` and `<main>` fragments on every visit and then we are going to leverage `data-pjax` attributes to replace an additional fragment when we navigate to the contact page.

### JavaScript Bundle

<!-- prettier-ignore -->
```javascript
import * as pjax from "@brixtol/pjax";


pjax.connect({
  targets: [
    "nav",
    "main"
  ]
});


```

### Home Page

Below we have a very basic Home Page with pjax wired up. All `<a>` elements will be intercepted and cached as per the default configuration. SSR web applications (in most cases) will only ever have a couple of fragments that change between navigation.

<!-- prettier-ignore -->
```html

<!doctype html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>Home Page</title>
    <script src="/bundle.js"></script>
  </head>
  <body>

    <header>

      <h1>@brixtol/pjax</h1>

      <!-- THIS FRAGMENT WILL BE REPLACED -->
      <nav>

        <!-- This link will be intercepted -->
        <a
         href="/home"
         class="active">Home</a>

        <!-- These links will be pe-fetched on hover -->
        <a
         href="/about"
         data-pjax-hover>About</a>

        <!-- This link will replace the #foo fragment -->
        <a
         href="/contact"
         data-pjax-replace="(['#foo'])">Contact</a>
      </nav>

    </header>

    <!-- THIS FRAGMENT WILL BE REPLACED -->
    <main>

      <h1>Welcome to the home page</h1>

      <div class="container">
        Brixtol Textiles is a Swedish apparel brand!
      </div>

    </main>

    <div id="foo">
      This fragment will not be touched until /contact is clicked
    </div>

    <footer>
      This fragment will not be touched during navigations as it's not
      defined within targets.
    </footer>

  </body>
</html>

```

### About Page

<br>

The about page in our web application would look practically identical to the home page. We instructed pjax to pre-fetch this page upon hover by annotating the `<a>` href link with `data-pjax-hover` attribute. This attribute informs pjax to being fetching the page the moment the user hovers over the `<a>` link which results in the visit being instantaneous. The **about** page only has some minor differences, but for the sake of clarity, lets have look:

<!-- prettier-ignore -->
```html

<!doctype html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>About Page</title>
    <script src="/bundle.js"></script>
  </head>
  <body>

    <header>

      <h1>@brixtol/pjax</h1>

      <!-- THIS FRAGMENT WILL BE REPLACED -->
      <nav>

        <!-- This link will be intercepted -->
        <a
         href="/home">Home</a>

        <!-- These links will be pe-fetched on hover -->
        <a
         href="/about"
         class="active"
         data-pjax-hover>About</a>

        <!-- This link will replace the #foo fragment -->
        <a
         href="/contact"
         data-pjax-replace="(['#foo'])">Contact</a>
      </nav>

    </header>

    <!-- THIS FRAGMENT WILL BE REPLACED -->
    <main>

      <h1>Welcome to the About Page</h1>

      <div class="container">
        Brixtol Textiles makes jackets out of recycled PET bottles.
        <p>Producing clothing in a sustainable way is the future!</p>
      </div>

    </main>

    <div id="foo">
      This fragment will not be touched until /contact is clicked
    </div>

    <footer>
      This fragment will not be touched during navigations as it's not
      defined within targets.
    </footer>

  </body>
</html>
```

### Contact Page

The contact page will replace an additional fragment with the id value of `foo` which we informed upon via attribute annotation. Upon visiting this page the `<nav>`, `<main>` and `<div id="foo">` fragments will be replaced.

<!-- prettier-ignore -->
```html

<!doctype html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>Contact Page</title>
    <script src="/bundle.js"></script>
  </head>
  <body>

    <header>

      <h1>@brixtol/pjax</h1>

      <!-- THIS FRAGMENT WILL BE REPLACED -->
      <nav>

        <!-- This link will be intercepted -->
        <a
         href="/home">Home</a>

        <!-- These links will be pe-fetched on hover -->
        <a
         href="/about"
         class="active"
         data-pjax-hover>About</a>

        <!-- This link will replace the #foo fragment -->
        <a
         href="/contact"
         data-pjax-replace="(['#foo'])">Contact</a>
      </nav>

    </header>

    <!-- THIS FRAGMENT WILL BE REPLACED -->
    <main>

      <h1>Welcome to the Contact Page</h1>

      <div class="container">
        This is contact page of our example! The below fragment was replaced too!
      </div>

    </main>

    <!-- THIS FRAGMENT WAS REPLACE VIA ATTRIBUTE INSTRUCTION -->
    <div id="foo">
      This fragment was replaced!
    </div>

    <footer>
      This fragment will not be touched during navigations as it's not
      defined within targets.
    </footer>

  </body>
</html>
```

</details>

# Lifecycle Events

Lifecycle events are dispatched to the document upon each navigation. You can access contextual information in the parameters. You can also cancel events with `preventDefault()` or by returning boolean `false` if you wish to prevent execution from occurring in a certain lifecycle.

The Pjax lifecycle events are dispatched in the following order of execution:

1. **pjax:ready**
2. **pjax:prefetch**
3. **pjax:trigger**
4. **pjax:request**
5. **pjax:cache**
6. **pjax:render**
7. **pjax:hydrate**
8. **pjax:module**
9. **pjax:load**

> When a hydrate event is triggered the `hydrate` method will be fired instead of the `render`.

### Events

<!-- prettier-ignore -->
```typescript
import type { Events } from '@brixtol/pjax'

// Triggered once after pjax initialization
pjax.on('connected', (state: IPage) => void)

// Triggered when a prefetch is triggered
pjax.on('prefetch', (target: Element, state: IPage, type: EventType) => void | false)

// Triggered when a mousedown event occurs on a link
pjax.on('trigger', (event: Event, state: IPage) => void | false)

// Triggered before a page is fetched over XHR
pjax.on('request', (state: IPage, type: EventType) => void | false)

// Triggered before a page and snapshot is saved to memory
pjax.on('cache', (state: IPage, snapshot: Document) => void | false | Document)

// Triggered on before a hydration replacement occurs
pjax.on('hydrate', (target: Element, newTarget: Element) => void | false)

// Triggered before a page or fragment is rendered
pjax.on('render', (target: Element, newTarget: Element) => void | false)

// Triggered after a page has rendered
pjax.on('load', (state: IPage) => void)

```

### `connected`

The connected event will be triggered after Pjax has connected and fired only once. This is the equivalent of the `DOMContentLoaded` event. Upon connection, Pjax will save the current documents outer HTML to the snapshot cache using `document.documentElement.outerHTML` whereas all additional snapshots are saved after an XHR request completes.

Because the initial snapshot is saved using `document.documentElement.outerHTML` the captured HTML may cause third party scripts which have augmented the document to serve an invalid dom into the snapshot cache. When a return navigation to this location occurs it may cause the third party script to fail. You can prevent issues of this nature from happening by initializing your modules within the `connected` event.

**Cancellable:** `false`

### `prefetch`

The prefetch event will be triggered for every prefetch request. Prefetch requests are fired when `hover`, `intersect` and `proximity` are triggered. This event will be frequently triggered if you are leveraging any of those capabilities. You can determine the type of prefetch which has occurred via the `type` parameter.

**Cancellable:** `true`

### `trigger`

The trigger event will be triggered when a `mousedown` event has occurred on a Pjax enabled `href`. This is the equivalent of a `click` and allows you to hook into the dispatch. Use this event to teardown any third party scripts.

**Cancellable:** `true`

### `request`

The request event will be triggered before an XHR request begins and a page is fetched. This event will be fired for `prefetch`, `hydrate` and `trigger` actions. You can determine the trigger action for the request using the `type` property passed in the `event.detail` parameter.

**Cancellable:** `true`

### `cache`

The cache event will be triggered immediately after a request has finished and before the snapshot and page record is saved to memory. You can determine the trigger action for the request via the `type` parameter. This Lifecycle also allows you to augment the snapshot `Document` and before it is saved.

**Cancellable:** `true`

### `render`

The render event will be triggered before a page or fragment is rendered (replaced) in the dom. For every `target` you've defined this event will fire. You can determine which elements are being replaced via the `target` and `newTarget` parameters passed. The `target` property represents the current element that will be replaced and the `newTarget` element represents the new target which it will be replaced with.

**Cancellable:** `true`

### `hydrate`

The hydrate event is identical to the `render` event. The parameters represent the current `target` and `newTarget` elements which will be replaced.

**Cancellable:** `true`

### `load`

The load event is the final lifecycle event to be triggered. Use this event to re-initialize any third party scripts. The load event will only execute after navigation has concluded.

**Cancellable:** `false`

# Methods

In addition to Lifecycle events, you also have a list of methods available. Methods will allow you some basic programmatic control of the Pjax session occurring, provides access to the cache store and various other operational utilities.

```typescript
import * as pjax from '@brixtol/pjax'

// Check to see if Pjax is supported by the browser
pjax.supported: boolean

// Connects Pjax, call to initialize a pjax session
pjax.connect(options?): void

// Returns the session model
pjax.session(): ISession

// Trigger hydration, optionally pass search params
pjax.hydrate(options: { url?: string, target: Elements[] }): Promise<IPage>

// Triggers a programmatic fetch (the record is not cached)
pjax.fetch(url: string): Promise<Document>

// Triggers a programmatic pre-fetch visit
pjax.prefetch(link: string | Element): Promise<IPage>

// Execute a programmatic pjax visit
pjax.visit(url?: string, options?): Promise<IPage>

// Access the cache, pass in href for specific record
pjax.cache(url?: string): Page{}

// Returns a snapshot or list of snapshots
pjax.snapshot(url?: string, options?): Document | { [id: string]: string }

// Clears the cache, pass in url to clear specific record
pjax.clear(url?: string): void

// Reloads the current page
pjax.reload(): Page{}

// Disconnects Pjax, ie: destroy the instance
pjax.disconnect(): void

```

# Attributes

Elements can be annotated with `data-pjax-*` attributes which you can leverage to customize how visits are rendered between navigations. By default, attributes are using a `-pjax-` schema but you can optionally provide a custom schema upon `connect` via the `schema` option.

- [data-pjax-eval](#data-pjax-eval)
- [data-pjax-disable](#data-pjax-disable)
- [data-pjax-hydrate](#data-pjax-hydrate)
- [data-pjax-replace](#data-pjax-replace)
- [data-pjax-prepend](#data-pjax-prepend)
- [data-pjax-append](#data-pjax-append)
- [data-pjax-hover](#data-pjax-hover)
- [data-pjax-threshold](#data-pjax-threshold)
- [data-pjax-proximity](#data-pjax-proximity)
- [data-pjax-position](#data-pjax-position)
- [data-pjax-cache](#data-pjax-cache)
- [data-pjax-history](#data-pjax-history)

## data-pjax-eval

Used on resources contained within the `<head>` fragment like styles, scripts or meta tags. Use this attribute if you want pjax to evaluate scripts and/or stylesheets. This option accepts a `false` value so you can define which scripts to execute per navigation. By default, pjax will run and evaluate all `<script>` tags it detects for every page visit but will not re-evaluate `<script src="*"></script>` tags.

> When a `<script>` tag is detected on a pjax navigation and annotated with `data-pjax-eval="false"` then execution will be triggered only once but never again after that.

<details>
<summary>
<strong>Tags</strong>
</summary>

The `data-pjax-eval` attribute can be annotated on any of the below HTML tags:

- `<meta>`
- `<link>`
- `<style>`
- `<script>`

</details>

<details>
<summary>
<strong>Values</strong>
</summary>

This attribute is a `boolean` type. Passing the `true` value is optional as `data-pjax-eval` infers truthy.

- `true`
- `false`

</details>

<details>
<summary>
<strong>Example</strong>
</summary>

```html
<script>
  console.log('I will run on every navigation');
</script>

<!-- script will only run once when between pjax visits -->
<script data-pjax-eval="false">
  console.log('I will run on initialization only!');
</script>
```

</details>

## data-pjax-disable

Use on `<a>` elements to disable pjax navigation. When a link element is annotated with `data-pjax-disable` a normal page navigation will be executed and cache will be cleared. You can optionally restore the cache using the `data-pjax-cache="restore"` attribute when navigating back to a pjax enabled url.

<details>
<summary>
<strong>Tags</strong>
</summary>

The `data-pjax-disable` attribute can be used on the following tags:

- `<a>`

</details>

<details>
<summary>
<strong>Values</strong>
</summary>

This attribute is a `truthy` type. Passing the `true` value is optional as `data-pjax-disable` infers truthy.

- `true`

</details>

<details>
<summary>
<strong>Examples</strong>
</summary>

Clicking this link will clear cache and a normal page navigation will be executed.

```html
<a href="*" data-pjax-disable></a>
```

Clicking this link will execute a normal page navigation but will inform pjax to save the current cache and restore it upon the next pjax enabled visit. See [data-pjax-cache](#data-pjax-cache) for more information.

```html
<a href="*" data-pjax-cache="restore" data-pjax-disable></a>
```

</details>

## data-pjax-track

Place on elements to track on a per-page basis that might otherwise not be contained within target elements. This methods will append nodes to the document `<body>` and placement is not certain. This is helpful when you need to reference an `svg` sprite of template or some sort.

<details>
<summary>
<strong>Tags</strong>
</summary>

The `data-pjax-track` attribute can be annotated on any HTML contained within `<body>` but cannot be applied to `<a>` href links.

</details>

<details>
<summary>
<strong>Values</strong>
</summary>

This attribute is a `truthy` type. Passing the `true` value is optional as `data-pjax-track` infers truthy.

- `true`

</details>

<details>
<summary>
<strong>Example</strong>
</summary>


changelog.md
# CHANGELOG

### 09/04/2022 | v1.0.0-beta.1

This module has now moved to an major release beta candidate. Multiple breaking changes have been applied in this version. Several refactors and improvement and overall maturity. Let's get into it.

#### Smaller Size

This release is 7.9kb (gzipped). Previous versions exceeded 10kb so the module is now considerably smaller and faster.

#### New setup configuration

Previous version were using a rather complicated setup configuration model. This is breaking change and application need to update to the new config model.

```js
import * as pjax from '@brixtol/pjax';

pjax.connect({
  targets: ['body'],
  timeout: 30000,
  poll: 15,
  async: true,
  cache: true,
  reverse: true,
  limit: 25,
  proximity: {
    bounding: 0,
    threshold: 100
  },
  hover: {
    trigger: 'href',
    threshold: 100
  },
  intersect: {
    rootMargin: '',
    threshold: ''
  },
  progress: {
    minimum: 0.08,
    easing: 'linear',
    speed: 200,
    trickle: true,
    threshold: 500,
    trickleSpeed: 200
  }
});
```

#### Improved scroll tracking

The way in which scroll position was tracked between navigations in previous versions was a tad costly on performance. This update improves upon this resource heavy aspect and scrolling is recognizably more snappier as a result.

#### Pub/Sub Event Dispatching

In previous versions the module allowed methods to be hooked into via `document` event listeners. This approach is replaced with a emitter pub/sub event dispatching. This allows users to leverage the lifecycle hooks more efficiently. The `pjax.on()` methods are the new approach to lifecycle hooks.

**Old Way**

```js
// Old way
document.addEventListener('pjax:load', ({ detail: { ... }}))
```

**New Way**

```js
pjax.on('load', param => {});
```

#### Improved `<script>` tag evaluation

A new and improved approach for handling inline `<script></script>` and external `<script src="*"></script>` javascript.

#### Added `schema` option

You can now customize the attribute annotation schema.

##### Exposed snapshot method

Provided an access point to snapshots in cache. You can return the `Map` holding all snapshots using `pjax.snapshot()` or optionally return a specific reference by passing `url` parameter (eg: `pjax.snapshot(url?: string)`).

#### Transit Purge

Hover pre-fetching sometimes delay the rendering speed between navigations with transit cache has multiple pending fetches. For example, in the Brixtol Textiles webshop we leverage hover pre-fetching on collection pages, which contain multiple urls. When product images are hovered pjax goes about fetching and saving a snapshot of each url to cache but if requests have not completed and a visitor clicks a product while fetches are still in transit then the clicked product will be added to the end of the queue. This results in minor delays between visits, because the pending fetches need to conclude before new ones begin.

This version introduces a transit purge which will abort all requests in queue besides the clicked request. This will free-up the congestion and allow the navigation to proceed.

### 21/12/2021 | v0.3.0-beta.1

##### Internal Improvements

Changes to some of the internal logic, specifically how state management is handled. Prior to this release the _store_ which held reference to snapshots and page state was using the native `Map` method. While nothing wrong with this approach, it felt better to leverage objects for `key > value` mapping because we constantly query the store. This version introduces a new helper utility for this internal operation leveraging a `null` prototype object reference. This helps improve the time it takes to query the store between navigations shaving around 30 to 70 milliseconds off the time it takes between fetches and becomes rather noticeable when a large cache exists.

In addition to the new state logic, a refactor an overhaul of the store handler. Previous version the store was wrapped in a `class` but JavaScript classes are fucked. The store is now just a collection of boring old functions. This refactor shaved some kb's of the total size of the project.

##### Hydrate Method

This is a new render method that will override defined `targets` which is helpful when you want to update specific fragments (elements) in the dom without replacing the main target fragments. This method updates the trigger page cache reference.

##### Anticipatory Caching

This version provides a preemptive fetches. There are 2 additions exposed in the version, which need a little more work but can be used. A new options for **reverse** caching is available which will execute a preemptive fetch of the previous page in the history stack when no snapshot exists in cache. Snapshot cache is purged if a browser refresh occurs and when navigating backwards or pages will need to be re-fetched resulting in minor delays if a refresh was triggered between browsing. By default, the last known previous page in the history stack is now fetched.

In addition to the reverse cache feature, a **preemptive** option is now available to the `prefetch` config. This option currently accepts a string list of paths only. Values defined here will be fetched preemptively and saved to cache either upon initial load and those pages will be visited in background asynchronously in the order they were passed. In the next release preemptive fetches will be made possible according to a specific path.

##### Methods

In previous version of the beta some exposed methods were not available or working properly. This release fixes them and introduces a new `Pjax.reload()` method which will trigger a fetch and re-cache of a page, while preserving scroll position. Some other fixes and updates made in this area include:

- The `Pjax.cache(url?:string)` method will return page cache
- The `Pjax.visit(url: string, state?: IPage)` method can digest `hydrate`

src/app/state.ts
import { IConfig, IPage, ISelectors } from 'types';
import { create } from '../constants/native';

/**
 * Configuration
 *
 * Initialization settings applied upon Pjax connection.
 * These are instance options, informing upon how the
 * pjax instance should run. The options defined here are
 * the defaults applied at runtime.
 */
export const config: IConfig = {
  targets: [ 'body' ],
  timeout: 30000,
  poll: 15,
  schema: 'pjax',
  async: true,
  cache: true,
  reverse: true,
  limit: 50,
  preload: null,
  hover: {
    trigger: 'attribute',
    threshold: 250
  },
  intersect: {
    rootMargin: '0px 0px 0px 0px',
    threshold: 0
  },
  proximity: {
    distance: 75,
    threshold: 250,
    throttle: 500
  },
  progress: {
    minimum: 0.08,
    easing: 'linear',
    speed: 200,
    trickle: true,
    threshold: 500,
    trickleSpeed: 200
  }
};

/**
 * Connects
 *
 * Determines the connection of various observers
 * and logic required for the Pjax instance.
 *
 * 1. Controller
 * 2. History
 * 3. Href
 * 4. Hover
 * 5. Intersect
 * 6. Scroll
 * 7. Proximity
 * 8. Head
 */
export const connect: Set<1 | 2 | 3 | 4 | 5 | 6 | 7 | 8> = new Set();

/**
 * Selectors
 *
 * String `key > value` references to DOM attributes
 * selectors used in the Pjax instance.
 */
export const schema: ISelectors = create(null);

/**
 * Position
 *
 * Holds the current position page offset radius.
 * The scroll position is updated and saved here.
 */
export const position: { x: number; y: number } = create(null);

/**
 * Pages
 *
 * Per-page state models. Each page uses a reference
 * configuration model. The objects in this store are
 * also available to the history API.
 */
export const pages: { [url: string]: IPage } = create(null);

/**
 * Snapshots
 *
 * This object holds documents responses of every page.
 * Each document is stored in string type. The key values
 * are unique ids generated using nanoid.
 */
export const snaps:{ [id: string]: string } = create(null);

/**
 * Events Model
 *
 * Holds an object reference for every event
 * emitted. Used by the event emitter operations
 */
export const events: { [name: string]: Array<() => void | boolean> } = create(null);

/**
 * Request Transits
 *
 * This object holds the XHR requests in transit. The object
 * properties represent the the request URL and the
 * value is the XML Request instance.
 */
export const transit: { [url: string]: XMLHttpRequest } = create(null);

/**
 * Tracked Elements
 *
 * Keeps a reference of tracked nodes between renders
 * and navigations to prevent extra appends from occuring.
 */
export const tracked: Set<string> = new Set();

/**
 * Request Timeouts
 *
 * Transit timers used to keep track of promises
 * and trigger operations like hover or proximity
 * prefetching.
 */
export const timers: { [url: string]: NodeJS.Timeout } = create(null);
